### PR TITLE
Meta unit

### LINKED ISSUE #1139: Create meta unit component
## Type
- ✨ Feature

## UI Requirements
screen shot 2017-07-10 at 10 54 34

## Acceptance Criteria
- [x] it should have a Helvetica Neue medium heading
- [x] it should accommodate free text (Helvetica Neue light), a list of links, a date or a date range
- [x] it should be underlined with a grey key-line

The design is (virtually) identical to what we currently have built (but not implemented) as a 'tags' component. This should almost certainly replace that component.

## Problem to be solved / User Story
In order to
 _learn more about an item_

As a
 _user_

I need/want/expect 
_to be able to read clearly labeled blocks of meta information about the item_

### PR DESCRIPTION
**Pull request by gestchild**
Closes #1139

## Type
✨ Feature  

## Value
Replaces tags component with met unit component for use on image item page

### CODE REVIEW

File: client/scss/utilities/_root-scope-classes.scss [hunk 1]
```diff
@@ -40,6 +40,14 @@
...
 }
 
+@each $color-key, $color-value in $colors {
+  .font-hover--#{$color-key}:hover {
+    // stylelint-disable sh-waqar/declaration-use-variable
+    color: nth($color-value, 1);
+    // stylelint-enable sh-waqar/declaration-use-variable
+  }
+}
+
```
Comment (jamesgorrie) [inline client/scss/utilities/_root-scope-classes.scss]: Do we have anything watching the size of our CSS compiled as we have a few of these now, which are less code, but more file?

File: client/scss/utilities/_root-scope-classes.scss [hunk 2]
```diff
@@ -92,6 +100,14 @@
   text-decoration: none;
 }
 
+.plain-list {
+  @include plain-list();
```
Comment (jamesgorrie) [inline client/scss/utilities/_root-scope-classes.scss]: Worth refactoring this out similar to the font things so we just have this on the elements that currently include it?

File: server/model/meta-unit.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// @flow
+import type {Link} from './link';
+import type {Time} from './time';
+
+export type MetaUnit = {|
+  headingLevel?: string;
```
Comment (jamesgorrie) [inline server/model/meta-unit.js]: number?

File: server/model/time.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export type Time = {|
+  text: string;
+  timestamp: string;
```
Comment (jamesgorrie) [inline server/model/time.js]: Date perhaps?

File: server/model/meta-unit.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  headingText: string;
+  content: Array<Link> | Array<Time> | string;
```
Comment (jamesgorrie) [inline server/model/meta-unit.js]: This seems quite heavily overloaded...

File: server/model/meta-unit.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+// @flow
+import type {Link} from './link';
+import type {Time} from './time';
+
+export type MetaUnit = {|
```
Comment (jamesgorrie) [inline server/model/meta-unit.js]: I can't quite figure out what a `MetaUnit` is?

File: server/model/meta-unit.js (hunk 3 above)
Comment (jamesgorrie) [inline server/model/meta-unit.js]: If we include the `h` - I would be tempted to call it `headingTag`.

File: server/views/components/meta-unit/meta-unit.config.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+export const handle = 'meta-unit';
+export const collated = true;
+
+export const metaUnit = createMetaUnit(({
+  type: 'text',
+  headingLevel: 'h3',
+  headingText: 'Curabitur quis',
+  content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur quis mollis turpis, eu facilisis ex. Vivamus sollicitudin pulvinar arcu sagittis posuere. Fusce a congue mauris. Vestibulum faucibus vel nibh ut commodo.',
+  includeDivider: true
+}: MetaUnit));
```
Comment (jamesgorrie) [inline server/views/components/meta-unit/meta-unit.config.js]: We're using createMetaUnit and enforcing the type here, which is superfluous.
I personally prefer type casting i.e. `({/*stuff...*/}: MetaUnit)`, but either works.

File: server/views/components/meta-unit/meta-unit.config.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+export const context = { model: metaUnit };
+
+export const variants = [
+  {
+    name: 'date',
+    context: {model: Object.assign({}, metaUnit, {type: 'date', content: [{text: '1731', datetime: '1731'}, {text: '1743', datetime: '1743'}]})}
```
Comment (jamesgorrie) [inline server/views/components/meta-unit/meta-unit.config.js]: We seem to be using `datetime` and `timestamp`?
I would be tempted to call it something, as it's not really either of these, it more a randomish datestring.

File: server/views/components/meta-unit/meta-unit.njk [hunk 9]
```diff
@@ -0,0 +1,35 @@
+<div>
+  <{{model.headingLevel or 'h2'}} class="
+    {{ {s:'HNM5' , m:'HNM4', l:'HNM4'} | fontClasses }}
```
Comment (jamesgorrie) [inline server/views/components/meta-unit/meta-unit.njk]: We could use `{{--}}` to save space.

File: server/views/components/meta-unit/meta-unit.njk [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    {{ {s:0, m:0, l:0, xl: 0} | spacingClasses({margin: ['top']}) }}
+    {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">
+   {{ model.headingText }}
+  </{{model.headingLevel or 'h2'}}>
+  {% if model.content | isString %}
+    <p class="{{ {s:'HNL5' , m:'HNL4', l:'HNL4', xl:'HNL4'} | fontClasses }} {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">{{model.content}}</p>
+  {% else %}
+    {% for item in model.content %}
+      {% if item.datetime %}
+        {% if loop.first %}
```
Comment (jamesgorrie) [inline server/views/components/meta-unit/meta-unit.njk]: We don't we just wrap the loop in the tag?
```twig
{% if model.content.length > 0 %}
<p>
  {% for item in model.content %}
</p>
{% .endif %}
```

It seems a little more readable.

File: server/views/components/meta-unit/meta-unit.njk [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+          <p class="{{ {s:'HNL5' , m:'HNL4', l:'HNL4', xl:'HNL4'} | fontClasses }} {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">
+        {% endif %}
+            <time datetime="{{ item.datetime }} ">{{ item.text }}</time>{{ ' - ' if not loop.last }}
+        {% if loop.last %}
+          </p>
+        {% endif %}
+      {% else %}
+          {% if loop.first %}
+          <ul class="plain-list {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">
```
Comment (jamesgorrie) [inline server/views/components/meta-unit/meta-unit.njk]: See my above question:
https://github.com/wellcometrust/wellcomecollection.org/pull/1166/files#r127148410

File: client/scss/utilities/_root-scope-classes.scss (hunk 1 above)
Comment (gestchild) [inline client/scss/utilities/_root-scope-classes.scss]: Have thought about that, it's pretty negligible especially with gzipping

File: client/scss/utilities/_root-scope-classes.scss (hunk 2 above)
Comment (gestchild) [inline client/scss/utilities/_root-scope-classes.scss]: yeah, there's a couple of other instances where things could be refactored like the text-meta class. I'll do them in another PR

File: server/model/meta-unit.js (hunk 3 above)
Comment (gestchild) [inline server/model/meta-unit.js]: will change to number

File: server/model/time.js (hunk 4 above)
Comment (gestchild) [inline server/model/time.js]: Did try that, but ran into issues - what it actually is, is a string with a specific format to be used in the <time> datetime attribute. Can we discuss this when you have a minute?

File: server/model/meta-unit.js (hunk 6 above)
Comment (gestchild) [inline server/model/meta-unit.js]: #1139 - does this help?

File: server/views/components/meta-unit/meta-unit.config.js (hunk 8 above)
Comment (gestchild) [inline server/views/components/meta-unit/meta-unit.config.js]: renamed and must have missed one.

it's for use in the datetime attribute of the time element, which requires is to be a string using ISO8601 standard

File: server/views/components/meta-unit/meta-unit.njk (hunk 10 above)
Comment (gestchild) [inline server/views/components/meta-unit/meta-unit.njk]: But don't want <p> on every instance of the loop

File: client/scss/utilities/_mixins.scss [hunk 12]
```diff
@@ -127,7 +127,7 @@
   }
 }
 
-// Allow an element to be horizontally stretched outwith its container
+// Allow an element to be horizontally stretched out with its container
```
Comment (davidpmccormick) [inline client/scss/utilities/_mixins.scss]: 'Outwith', one word, was what I'd meant, but I realise it's [a Scottish-ism](https://www.google.co.uk/search?q=outwith&oq=outwith&aqs=chrome..69i57j0l5.3866j0j7&sourceid=chrome&ie=UTF-8). Maybe 'outside of' would be better?

File: server/filters/is-string.js [hunk 13]
```diff
@@ -0,0 +1,3 @@
+export function isString(string) {
```
Comment (davidpmccormick) [inline server/filters/is-string.js]: nitpick – if w're checking whether or not it's a string, then the parameter name should probably be something less specific, like `input`, or `maybeString`

File: client/scss/utilities/_mixins.scss (hunk 12 above)
Comment (gestchild) [inline client/scss/utilities/_mixins.scss]: Ah, not heard of that before, will change to 'outside of' if ok, as I didn't get that, that's what outwith meant

File: server/views/components/meta-unit/meta-unit.njk [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+   {{ model.headingText }}
+  </{{model.headingLevel or 'h2'}}>
+  {% if model.content | isString %}
+    <p class="{{ {s:'HNL5' , m:'HNL4', l:'HNL4', xl:'HNL4'} | fontClasses }} {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">{{model.content}}</p>
+  {% else %}
+    {% for item in model.content %}
+      {% if item.datetime %}
+        {% if loop.first %}
+          <p class="{{ {s:'HNL5' , m:'HNL4', l:'HNL4', xl:'HNL4'} | fontClasses }} {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">
+        {% endif %}
```
Comment (davidpmccormick) [inline server/views/components/meta-unit/meta-unit.njk]: Indenting gets a bit off here

File: server/views/components/meta-unit/meta-unit.njk [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            <time datetime="{{ item.datetime }} ">{{ item.text }}</time>{{ ' - ' if not loop.last }}
+        {% if loop.last %}
+          </p>
+        {% endif %}
+      {% else %}
+          {% if loop.first %}
+          <ul class="plain-list {{ {s:1, m:1, l:1, xl: 1} | spacingClasses({margin: ['bottom']}) }}">
+          {% endif %}
+            <li class="inline"><a class="plain-link font--elf-green font-hover--mint {{ {s:'HNM5' , m:'HNM4', l:'HNM4', xl: 'HNM4'} | fontClasses }}" href="{{ item.url }}">{{ item.text }}</a>{{ ', ' if not loop.last }}</li>
```
Comment (davidpmccormick) [inline server/views/components/meta-unit/meta-unit.njk]: `font--elf-green font-hover--mint` – almost certainly the best class names I'm going to see this week :1st_place_medal:

**Pull request merged**